bpf/lib/nat.h

```diff
@@ -441,6 +441,52 @@ static __always_inline bool snat_v4_can_skip(const struct ipv4_nat_target *targe
 	return false;
 }
 
+static __always_inline int snat_v4_create_dsr(struct __sk_buff *skb,
+					      __be32 to_saddr, __be16 to_sport)
+{
+	void *data, *data_end;
+	struct ipv4_ct_tuple tuple = {};
+	struct ipv4_nat_entry state = {};
+	struct iphdr *ip4;
+	struct {
+		__be16 sport;
+		__be16 dport;
+	} l4hdr;
+	__u32 off;
+
+	build_bug_on(sizeof(struct ipv4_nat_entry) > 64);
+
+	if (!revalidate_data(skb, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+	tuple.nexthdr = ip4->protocol;
+	tuple.daddr = ip4->saddr;
+	tuple.saddr = ip4->daddr;
+	tuple.flags = NAT_DIR_EGRESS;
+	off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
+	switch (tuple.nexthdr) {
+	case IPPROTO_TCP:
+	case IPPROTO_UDP:
```

Comment:
What about ICMP which is handled partially in NAT engine?

Reply:
An ICMP packet cannot be sent to a NodePort service. I will add a comment.